distributed-process-tests/src/Control/Distributed/Process/Tests/Mx.hs

```diff
@@ -174,32 +199,159 @@ testAgentEventHandling result = do
   seenDead  <- receiveChan reply
 
   stash result $ seenAlive && seenDead
+  kill agentPid "test-complete" >> awaitExit agentPid
+  where
+    faff delay port = do
+      res <- receiveChanTimeout delay port
+      unless (isNothing res) $ do
+        pause delay
+        faff delay port
+
+testMxRegEvents :: TestResult () -> Process ()
+testMxRegEvents result = do
+  {- This test only deals with the local case, to ensure that we are being
+     notified in the expected order - the remote cases related to the
+     behaviour of the node controller are contained in the CH test suite. -}
+  ensure (stash result ()) $ do
+    let label = "testMxRegEvents"
+    let agentLabel = "mxRegEvents-agent"
+    let delay = 1000000
+    (regChan, regSink) <- newChan
+    (unRegChan, unRegSink) <- newChan
+    agent <- mxAgent (MxAgentId agentLabel) () [
+        mxSink $ \ev -> do
+          case ev of
+            MxRegistered pid label'
+              | label' == label -> liftMX $ sendChan regChan (label', pid)
+            MxUnRegistered pid label'
+              | label' == label -> liftMX $ sendChan unRegChan (label', pid)
+            _                   -> return ()
+          mxReady
+      ]
+
+    p1 <- spawnLocal expect
+    p2 <- spawnLocal expect
+
+    register label p1
+    reg1 <- receiveChanTimeout delay regSink
+    reg1 `shouldBe` equalTo (Just (label, p1))
+
+    unregister label
+    unreg1 <- receiveChanTimeout delay unRegSink
+    unreg1 `shouldBe` equalTo (Just (label, p1))
+
+    register label p2
+    reg2 <- receiveChanTimeout delay regSink
+    reg2 `shouldBe` equalTo (Just (label, p2))
+
+    reregister label p1
+    unreg2 <- receiveChanTimeout delay unRegSink
+    unreg2 `shouldBe` equalTo (Just (label, p2))
+
+    reg3 <- receiveChanTimeout delay regSink
+    reg3 `shouldBe` equalTo (Just (label, p1))
+
+    mapM_ (flip kill $ "test-complete") [agent, p1, p2]
+    awaitExit agent
+
+testMxRegMon :: LocalNode -> TestResult () -> Process ()
+testMxRegMon remoteNode result = do
+  ensure (stash result ()) $ do
+    -- ensure that when a registered process dies, we get a notification that
+    -- it has been unregistered as well as seeing the name get removed
+    let label1 = "aaaaa"
+    let label2 = "bbbbb"
+    let isValid l = l == label1 || l == label2
+    let agentLabel = "mxRegMon-agent"
+    let delay = 1000000
```

Comment:
it's a pity that we have to use `delay` but ok for now.

Reply:
Well, the real issue is that we can see events related to monitoring agents from other monitoring agents...

So one solution is to run each test case in a new local node. That might actually be a better approach here, tbh. Maybe I will re-cook the patch accordingly and resubmit... Or do that as part of the next patch I'm working on, which has management/tracing fixes and improvements?

Reply:
It's ok to have that in another PR.